src/Chainweb/Pact/PactService.hs

```diff
@@ -648,51 +672,51 @@ rewindTo mb act = do
 
     failNonGenesisOnEmptyDb = fail "impossible: playing non-genesis block to empty DB"
 
-    playFork cp bhdb payloadDb newH parentHash (lastBlockHeight, lastHash) =
-      flip catches exHandlers $ do
-          parentHeader <- liftIO $ lookupM bhdb parentHash
-          lastHeader <- liftIO $ lookupM bhdb lastHash
-          (!_, _, newBlocks) <-
-              liftIO $ collectForkBlocks bhdb lastHeader parentHeader
-          -- play fork blocks
-          V.mapM_ (fastForward payloadDb) newBlocks
-          -- play new block
-          restoreCheckpointer (Just (newH, parentHash)) >>= act
+    playFork cp bhdb payloadDb newH parentHash (lastBlockHeight, lastHash) = do
+        parentHeader <- liftIO $ lookupM bhdb parentHash
+        lastHeader <- findValidParent lastBlockHeight lastHash
+
+        (!_, _, newBlocks) <-
+            liftIO $ collectForkBlocks bhdb lastHeader parentHeader
+        -- play fork blocks
+        V.mapM_ (fastForward payloadDb) newBlocks
+        -- play new block
+        restoreCheckpointer (Just (newH, parentHash))
       where
-        exHandlers = [Handler handleTreeDbFailure, Handler handlePayloadFailure]
-        handleEx :: forall e . Exception e => e -> PactServiceM cas a
-        handleEx e =
-              (liftIO $ getBlockParent cp (lastBlockHeight, lastHash)) >>= \case
-                Nothing -> throwM e
-                Just hash -> playFork cp bhdb payloadDb newH parentHash
-                                      (pred lastBlockHeight, hash)
-        handleTreeDbFailure e@(_ :: TreeDbException BlockHeaderDb) = handleEx e
-        handlePayloadFailure e@(_ :: PayloadNotFoundException) = handleEx e
+        findValidParent height hash = liftIO (lookup bhdb hash) >>= \case
```

Comment:
How could have the previous code bitten us in the butt?

Reply:
I guess it never threw an exception. If it did, the node probably would have run out of memory and crashed. Maybe it happened, and we didn't notice what was the reason.